### PR TITLE
fix: default values for Session Redis Handler

### LINKED ISSUE #6512: Bug: Redis::connect(): Passing null to parameter #3 ($timeout) of type float is deprecated
### PHP Version

8.1

### CodeIgniter4 Version

4.2.6

### CodeIgniter4 Installation Method

Composer (as dependency to an existing project)

### Which operating systems have you tested for this bug?

macOS

### Which server did you use?

fpm-fcgi

### Database

_No response_

### What happened?

将session 设置redis

php7.4正常

php8.1报错：Redis::connect(): Passing null to parameter #3 ($timeout) of type float is deprecated

image

### Steps to Reproduce

将session 设置redis

php8.1 下

### Expected Output

将session 设置redis

php8.1 下

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by kenjis**
**Description**
Fixes  #6512

Make the same as Cache Redis config.
https://github.com/codeigniter4/CodeIgniter4/blob/ed0cdd00eaf20f7d6efb373af991111ead4fece8/app/Config/Cache.php#L155-L161

Ref, https://github.com/phpredis/phpredis#php-session-handler

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: system/Session/Handlers/RedisHandler.php [hunk 1]
```diff
@@ -75,10 +77,10 @@ public function __construct(AppConfig $config, string $ipAddress)
 
             $this->savePath = [
                 'host'     => $matches[1],
-                'port'     => empty($matches[2]) ? null : $matches[2],
+                'port'     => empty($matches[2]) ? 6379 : $matches[2],
```
Comment (MGatner) [inline system/Session/Handlers/RedisHandler.php:80]: I recommend a class constant instead of the number.

**Review by MGatner — COMMENTED** [review]
Really great test case! One change to consider.

File: system/Session/Handlers/RedisHandler.php (hunk 1 above)
Comment (kenjis) [reply system/Session/Handlers/RedisHandler.php:80]: I just set the default port number for Redis, but I think again,
It seems odd that `tcp://127.0.0.1` means `tcp://127.0.0.1:6379`?
Is it better to raise an error?

Comment (MGatner) [reply system/Session/Handlers/RedisHandler.php:80]: I don't see a problem with that. It looks like it is a common setup for local Redis dev: https://duckduckgo.com/?q=+tcp%3A%2F%2F127.0.0.1%3A6379&t=iphone&ia=qa

Comment (kenjis) [reply system/Session/Handlers/RedisHandler.php:80]: `tcp://127.0.0.1:6379` is no problem.
But does `tcp://127.0.0.1` equals `tcp://127.0.0.1:6379`?
I doubt it.

Comment (MGatner) [reply system/Session/Handlers/RedisHandler.php:80]: That, my friend, is beyond me. It looks like the port is pre-specified in the Config file (https://raw.githubusercontent.com/redis/redis/7.0/redis.conf) so my hunch is "yes", but without testing I don't know.

File: system/Session/Handlers/RedisHandler.php [hunk 2]
```diff
@@ -75,10 +77,10 @@ public function __construct(AppConfig $config, string $ipAddress)
...
                 'password' => preg_match('#auth=([^\s&]+)#', $matches[3], $match) ? $match[1] : null,
-                'database' => preg_match('#database=(\d+)#', $matches[3], $match) ? (int) $match[1] : null,
-                'timeout'  => preg_match('#timeout=(\d+\.\d+)#', $matches[3], $match) ? (float) $match[1] : null,
+                'database' => preg_match('#database=(\d+)#', $matches[3], $match) ? (int) $match[1] : 0,
+                'timeout'  => preg_match('#timeout=(\d+\.\d+)#', $matches[3], $match) ? (float) $match[1] : 0,
```
Comment (paulbalandan) [inline system/Session/Handlers/RedisHandler.php:83]: Should this be `0.0`?

Comment (paulbalandan) [reply system/Session/Handlers/RedisHandler.php:83]: I think the regex should also be updated to check for int timeouts.

File: tests/system/Session/Handlers/Database/RedisHandlerTest.php [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+{
+    private string $sessionName     = 'ci_session';
+    private string $sessionSavePath = 'tcp://127.0.0.1:6379';
+    private string $userIpAddress   = '127.0.0.1';
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        if (! class_exists(Redis::class)) {
```
Comment (paulbalandan) [inline tests/system/Session/Handlers/Database/RedisHandlerTest.php:32]: How about moving this to setUpBeforeClass? Or, using a class docblock instead, like `@requires extension redis`?

File: system/Session/Handlers/RedisHandler.php (hunk 1 above)
Comment (kenjis) [reply system/Session/Handlers/RedisHandler.php:80]: I checked the test in this PR with `$sessionSavePath = 'tcp://127.0.0.1'`on PHP 8.0,
and the tests passed.
So to keep compatibility we need to accept `tcp://127.0.0.1` as `tcp://127.0.0.1:6379` in this case.

Comment (kenjis) [reply system/Session/Handlers/RedisHandler.php]: Refactored to use class constant.

File: system/Session/Handlers/RedisHandler.php (hunk 2 above)
Comment (kenjis) [reply system/Session/Handlers/RedisHandler.php]: Done.

File: tests/system/Session/Handlers/Database/RedisHandlerTest.php (hunk 3 above)
Comment (kenjis) [reply tests/system/Session/Handlers/Database/RedisHandlerTest.php]: Oh, thanks. It is better.
Done.

**Review by MGatner — APPROVED** [review]
👌

File: system/Session/Handlers/RedisHandler.php [hunk 4]
```diff
@@ -22,6 +22,8 @@
  */
 class RedisHandler extends BaseHandler
 {
+    private const REDIS_DEFAULT_PORT = 6379;
```
Comment (mostafakhudair) [inline system/Session/Handlers/RedisHandler.php:25]:
```suggestion
    private const DEFAULT_PORT = 6379;
```

What about `RedisHandler::DEFAULT_PORT`?

Comment (kenjis) [reply system/Session/Handlers/RedisHandler.php]: Yeah, `REDIS_` is not needed.
Done.

**Review by datamweb — APPROVED** [review]

**Review by MGatner — APPROVED** [review]

**Pull request merged**